keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1179 @@
+
+Here are some common change scenarios and whether the change needs to be preserved or not:
+* API change [Yes]
+* Change of supported systems [Yes]
+* Bug fix [No]
+* Performance optimizations [No]
+* Unstable metrics change [No]
+* Code refactoring [No]
+
+Listed below are some concrete examples of feature changes:
+**Feature**|**Changes That Should Be Preserved**|**Changes That Do Not Need To Be Preserved**
+-----|-----|-----
+APIPriorityAndFairness | [add v1beta3 for Priority And Fairness](https://github.com/kubernetes/kubernetes/pull/112306) | [More seat metrics for APF](https://github.com/kubernetes/kubernetes/pull/105873)
+ValidatingAdmissionPolicy | | [Drop AvailableResources from controller context](https://github.com/kubernetes/kubernetes/pull/117977), [Encapsulate KCM controllers with their metadata](https://github.com/kubernetes/kubernetes/pull/120371)
+APIServerTracing | | [Revert "Graduate API Server tracing to beta"](https://github.com/kubernetes/kubernetes/pull/113803)
+MemoryManager | | [Don't reuse memory of a restartable init container](https://github.com/kubernetes/kubernetes/pull/120715)
+NodeSwap | if done after promoting to beta: [Add full cgroup v2 swap support and remove cgroup v1 support](https://github.com/kubernetes/kubernetes/pull/118764) | [only configure swap if swap is enabled](https://github.com/kubernetes/kubernetes/pull/120784)
+
+To preserve the behavior, naively the feature implementations can be gated by version number. 
+For example, if `FeatureA` is partially implemented in 1.28 and additional functionality
+is added in 1.29, the feature developer is expected to gate the functionality by version.
+E.g.:
+
+```go
+if feature_gate.Enabled(FeatureA) && feature_gate.CompatibilityVersion() <= "1.28" {implementation 1}
+if feature_gate.Enabled(FeatureA) && feature_gate.CompatibilityVersion() >= "1.29" {implementation 2}
+```
+
+A better way might be to define a `featureOptions` struct constructed based on the the feature gate, and have the `featureOptions` control the main code flow, so that the main code is version agnostic. 
+E.g.:
+```go
+// in kube_features.go
+const (
+  FeatureA featuregate.Feature = "FeatureA"
+  FeatureB featuregate.Feature = "FeatureB"
+)
+
+func init() {
+	utilfeature.DefaultMutableFeatureGate.AddVersioned(defaultKubernetesFeatureGates)
+}
+
+var defaultKubernetesFeatureGates = map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.27"), Default: false, PreRelease: Alpha},
+		},
+    featureB: VersionedSpecs{
+			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.30"), Default: true, PreRelease: Beta},
+		},
+}
+
+type featureOptions struct {
+  AEnabled        bool
+  AHasCapabilityZ bool
+  BEnabled        bool
+  BHandler        func()
+}
```

Comment:
This seemingly promotes a "features-of-features" paradigm for feature authors to consider. I get why it's necessary but I am ehhhh about this idea.

Am I correct in assuming most features will not need to do anything like this? I would expect most "breaking" changes to be added in alpha.

Reply:
Yes, this would be very rare. You can think of it as: if you change some important design of your feature that has been broadly adopted, you have 2 choices: introducing a new feature, or if you want to keep using the same feature name, these are the hoops you have to jump through.